pkg/restore/cdclog/buffer.go

```diff
@@ -0,0 +1,165 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cdclog
+
+import (
+	"context"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/types"
+	"go.uber.org/zap"
+)
+
+// TableBuffer represents the kv buffer of this table.
+// we restore one tableBuffer in one goroutine.
+// this is the concurrent unit of log restore.
+type TableBuffer struct {
+	KvPairs []Row
+	Size    int64
+
+	KvEncoder Encoder
+	tableInfo table.Table
+
+	flushKVPairs int
+
+	colNames []string
+	colPerm  []int
+}
+
+// NewTableBuffer creates TableBuffer.
+func NewTableBuffer(tbl table.Table, flushKVPairs int) *TableBuffer {
+	kvEncoder := NewTableKVEncoder(tbl, &SessionOptions{
+		Timestamp: time.Now().Unix(),
+		// TODO make it config
+		RowFormatVersion: "1",
+	})
+
+	columns := tbl.Meta().Cols()
+	colNames := make([]string, 0, len(columns))
+	colPerm := make([]int, 0, len(columns)+1)
+
+	for i, col := range columns {
+		colNames = append(colNames, col.Name.String())
+		colPerm = append(colPerm, i)
+	}
+	if TableHasAutoRowID(tbl.Meta()) {
+		colPerm = append(colPerm, -1)
+	}
+
+	return &TableBuffer{
+		KvPairs:   make([]Row, 0, flushKVPairs),
+		KvEncoder: kvEncoder,
+		tableInfo: tbl,
+
+		flushKVPairs: flushKVPairs,
+
+		colNames: colNames,
+		colPerm:  colPerm,
+	}
+}
+
+func (t *TableBuffer) translateToDatum(row map[string]column) ([]types.Datum, error) {
+	cols := make([]types.Datum, 0, len(row))
+	for _, col := range t.colNames {
+		val, err := row[col].toDatum()
+		if err != nil {
+			return nil, err
+		}
+		log.Debug("translate to datum", zap.String("col", col), zap.Stringer("val", val))
+		cols = append(cols, val)
+	}
+	return cols, nil
+}
+
+// Append appends the item to this buffer.
+func (t *TableBuffer) Append(ctx context.Context, item *SortItem) error {
+	log.Debug("Append item to buffer",
+		zap.Stringer("table", t.tableInfo.Meta().Name),
+	)
+	row := item.Meta.(*MessageRow)
+
+	if row.Update != nil {
+		if row.PreColumns != nil {
+			log.Debug("process update event", zap.Any("row", row))
+			oldCols, err := t.translateToDatum(row.PreColumns)
+			if err != nil {
+				return err
+			}
+
+			pair, err := t.KvEncoder.RemoveRecord(oldCols, item.RowID, t.colPerm)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			t.KvPairs = append(t.KvPairs, pair)
+			t.Size++
+
+			newCols, err := t.translateToDatum(row.Update)
+			if err != nil {
+				return err
+			}
+			pair, err = t.KvEncoder.AddRecord(newCols, item.RowID, t.colPerm)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			t.KvPairs = append(t.KvPairs, pair)
+			t.Size++
+		} else {
```

Comment:
this code block is duplicated with the if block. The `update` code should be same for insesrt/update event. So we can check if it is update event, then add the deleted keys